src/disk_cache.rs

```diff
@@ -0,0 +1,952 @@
+        root_folder: std::path::PathBuf,
+        part_size: usize,
+    ) -> Self {
+        Self {
+            object_store,
+            root_folder,
+            part_size,
+        }
+    }
+
+    async fn cached_head(&self, location: &Path) -> object_store::Result<ObjectMeta> {
+        let entry = DiskCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size: self.part_size,
+        };
+        match entry.read_meta().await {
+            Ok(Some(meta)) => Ok(meta),
+            _ => self.object_store.head(location).await,
+        }
+    }
+
+    async fn cached_get_opts(
+        &self,
+        location: &Path,
+        opts: GetOptions,
+    ) -> object_store::Result<GetResult> {
+        let entry = DiskCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size: self.part_size,
+        };
+
+        let meta = self.maybe_prefetch_range(&entry, &opts.range).await?;
+        let range = self.canonicalize_range(opts.range, meta.size)?;
+        let parts = self.split_range_into_parts(range.clone());
+
+        // read parts, and concatenate them into a single stream. please note that some of these part may not be cached,
+        // we'll still fallback to the object store to get the missing parts.
+        let futures = parts
+            .into_iter()
+            .map(|(part_id, range_in_part)| self.read_part(location, part_id, range_in_part))
+            .collect::<Vec<_>>();
+        let result_stream = stream::iter(futures).then(|fut| fut).boxed();
+
+        Ok(GetResult {
+            meta,
+            range,
+            attributes: Default::default(),
+            payload: GetResultPayload::Stream(result_stream),
+        })
+    }
+
+    // if an object is not cached before, maybe_prefetch_range will try to prefetch the object from the
+    // object store and save the parts into the local disk cache. the prefetching is helpful to reduce the
+    // number of GET requests to the object store, it'll try to aggregate the parts among the range into a
+    // single GET request, and save the related parts into local disks together.
+    // when it sends GET requests to the object store, the range is expected to be ALIGNED with the part
+    // size.
+    async fn maybe_prefetch_range(
+        &self,
+        entry: &DiskCacheEntry,
+        range: &Option<GetRange>,
+    ) -> object_store::Result<ObjectMeta> {
+        match entry.read_meta().await {
+            Ok(Some(meta)) => return Ok(meta),
+            Ok(None) => {}
+            Err(_) => {
+                // TODO: add a warning log
+            }
+        };
+
+        let aligned_opts = match &range {
+            None => GetOptions::default(),
+            Some(get_range) => GetOptions {
+                range: Some(self.align_get_range(get_range)),
+                ..Default::default()
```

Comment:
i am afraid that having versions will make caches much harder, this cache store assumes that all the files are immutable, so that we can safely store caches without worring about cache invalidation.